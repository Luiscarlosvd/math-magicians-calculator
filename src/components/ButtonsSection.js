import React from 'react';
import './button.css';
import PropTypes from 'prop-types';

const ButtonsSection = ({ onClick }) => (
  <div className="buttons-calculator">
    <button type="button" className="button">AC</button>
    <button type="button" onClick={onClick} className="button">+/-</button>
    <button type="button" onClick={onClick} className="button">%</button>
    <button type="button" onClick={onClick} className="button button-orange">รท</button>
    <button type="button" onClick={onClick} className="button">7</button>
    <button type="button" onClick={onClick} className="button">8</button>
    <button type="button" onClick={onClick} className="button">9</button>
    <button type="button" onClick={onClick} className="button button-orange">X</button>
    <button type="button" onClick={onClick} className="button">4</button>
    <button type="button" onClick={onClick} className="button">5</button>
    <button type="button" onClick={onClick} className="button">6</button>
    <button type="button" onClick={onClick} className="button button-orange">-</button>
    <button type="button" onClick={onClick} className="button">1</button>
    <button type="button" onClick={onClick} className="button">2</button>
    <button type="button" onClick={onClick} className="button">3</button>
    <button type="button" onClick={onClick} className="button button-orange">+</button>
    <button type="button" onClick={onClick} className="button" id="button-0">0</button>
    <button type="button" onClick={onClick} className="button">.</button>
    <button type="button" className="button button-orange">=</button>
  </div>
);

ButtonsSection.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default ButtonsSection;
