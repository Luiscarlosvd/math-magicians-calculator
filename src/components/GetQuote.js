import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './quote.css';

const headers = {
  'X-Api-Key': 'VtmOQOFHRFLo4Lyp7G9bZw==7m0QdxrDfY5mRlsF',
};

const GetQuote = () => {
  const [quote, setQuote] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    axios.get('https://api.api-ninjas.com/v1/quotes?category=movies', { headers })
      .then((resp) => {
        setQuote(resp.data[0]);
        setIsLoading(false);
      }).catch(() => {
        setIsLoading(false);
        setError(true);
      });
  }, [headers]);
  return (
    <div className="quote">
      {isLoading ? (
        <p>Loading ...</p>
      ) : (
        <>
          <h3>{quote.quote}</h3>
          <p>{quote.author}</p>
        </>
      )}
      <>
        {error ? (
          <p>Sorry... we are experiencing some errors.</p>
        ) : (
          false
        )}
      </>
    </div>
  );
};

export default GetQuote;
