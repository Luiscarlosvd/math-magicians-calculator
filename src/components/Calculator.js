import React, { useState } from 'react';
import './calculator.css';
import ButtonsSection from './ButtonsSection';
import calculate from '../logic/calculate';

function Calculator() {
  const [result, setResult] = useState({
    total: null,
    next: 0,
    operation: null,
  });
  const handleValue = (e) => {
    setResult((prevState) => calculate(prevState, e.target.textContent));
  };
  const showResult = (total, next, operation) => {
    const invalid = [null, undefined];
    const showTotal = invalid.includes(total) ? '' : total;
    const showNext = invalid.includes(next) ? '' : next;
    const showOp = invalid.includes(operation) ? '' : operation;

    const show = `${showTotal}${showOp}${showNext}`;
    return show;
  };
  return (
    <div className="calculator-section-container">
      <h2>Let&apos;s do some Math</h2>
      <div className="calculator-container">
        <div className="div-result">{showResult(result.total, result.next, result.operation)}</div>
        <ButtonsSection onClick={handleValue} />
      </div>
    </div>
  );
}

export default Calculator;
