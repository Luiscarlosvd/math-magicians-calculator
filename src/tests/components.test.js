import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import axios from 'axios';
import ButtonsSection from '../components/ButtonsSection';
import Calculator from '../components/Calculator';
import GetQuote from '../components/GetQuote';

// ButtonSection component tests

describe('ButtonsSection', () => {
  test('should match the snapshot', () => {
    // Arrange
    const onClick = jest.fn();

    // Act
    const { container } = render(<ButtonsSection onClick={onClick} />);

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should call the onClick prop', () => {
    // Arrange
    const onClick = jest.fn();
    const { getByText } = render(<ButtonsSection onClick={onClick} />);

    // Act
    fireEvent.click(getByText('1'));

    // Assert
    expect(onClick).toHaveBeenCalled();
  });
});

// Calculator component test

describe('Calculator', () => {
  test('should match the snapshot', () => {
    // Arrange

    // Act
    const { container } = render(<Calculator />);

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should update the result with the correct value', () => {
    // Arrange
    const { getByText } = render(<Calculator />);
    const expectedValue = '3';

    // Act
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('+'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('='));
    const resultElement = document.querySelector('.div-result');
    
    // Assert
    expect(resultElement).toHaveTextContent(expectedValue);
  });
});

// GetQuote component tests

describe('GetQuote', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  })
  test('should display loading text on the screen when waiting for api response', () => {
    // Arrange
    jest.spyOn(axios, 'get').mockImplementation(() => new Promise(() => {}))
    // Act
    render(<GetQuote />);

    // Assert
    const loadingText = screen.getByText('Loading ...');
    expect(loadingText).toBeInTheDocument();
  });

  test('should display the quote on the screen when it gets the data from the Api', async () => {
    // Arrange
    const responseQuote = {
      data: [
        {
          quote: 'Hello',
          author: 'Luis',
        },
      ],
    };
    jest.spyOn(axios, 'get').mockResolvedValue(responseQuote)

    // Act
    render(<GetQuote />);

    // Assert
    const quoteText = await screen.findByText(`“Hello” - Luis`);
    expect(quoteText).toBeInTheDocument();
  });

  test('should display error message on the screen when api get method fails', async () => {
    // Arrange
    jest.spyOn(axios, 'get').mockRejectedValue(new Error('API request failed'));

    // Act
    render(<GetQuote />);

    // Assert
    const errorMessage = await screen.findByText('Sorry... We are experiencing some errors..');
    expect(errorMessage).toBeInTheDocument();
  });
});