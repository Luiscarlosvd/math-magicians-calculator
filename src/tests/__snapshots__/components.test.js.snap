// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonsSection should match the snapshot 1`] = `
<div
  class="buttons-calculator"
>
  <button
    class="button"
    type="button"
  >
    AC
  </button>
  <button
    class="button"
    type="button"
  >
    +/-
  </button>
  <button
    class="button"
    type="button"
  >
    %
  </button>
  <button
    class="button button-orange"
    type="button"
  >
    รท
  </button>
  <button
    class="button"
    type="button"
  >
    7
  </button>
  <button
    class="button"
    type="button"
  >
    8
  </button>
  <button
    class="button"
    type="button"
  >
    9
  </button>
  <button
    class="button button-orange"
    type="button"
  >
    x
  </button>
  <button
    class="button"
    type="button"
  >
    4
  </button>
  <button
    class="button"
    type="button"
  >
    5
  </button>
  <button
    class="button"
    type="button"
  >
    6
  </button>
  <button
    class="button button-orange"
    type="button"
  >
    -
  </button>
  <button
    class="button"
    type="button"
  >
    1
  </button>
  <button
    class="button"
    type="button"
  >
    2
  </button>
  <button
    class="button"
    type="button"
  >
    3
  </button>
  <button
    class="button button-orange"
    type="button"
  >
    +
  </button>
  <button
    class="button"
    id="button-0"
    type="button"
  >
    0
  </button>
  <button
    class="button"
    type="button"
  >
    .
  </button>
  <button
    class="button button-orange"
    type="button"
  >
    =
  </button>
</div>
`;

exports[`Calculator should match the snapshot 1`] = `
<div
  class="calculator-section-container"
>
  <h2>
    Let's do some Math
  </h2>
  <div
    class="calculator-container"
  >
    <div
      class="div-result"
    >
      0
    </div>
    <div
      class="buttons-calculator"
    >
      <button
        class="button"
        type="button"
      >
        AC
      </button>
      <button
        class="button"
        type="button"
      >
        +/-
      </button>
      <button
        class="button"
        type="button"
      >
        %
      </button>
      <button
        class="button button-orange"
        type="button"
      >
        รท
      </button>
      <button
        class="button"
        type="button"
      >
        7
      </button>
      <button
        class="button"
        type="button"
      >
        8
      </button>
      <button
        class="button"
        type="button"
      >
        9
      </button>
      <button
        class="button button-orange"
        type="button"
      >
        x
      </button>
      <button
        class="button"
        type="button"
      >
        4
      </button>
      <button
        class="button"
        type="button"
      >
        5
      </button>
      <button
        class="button"
        type="button"
      >
        6
      </button>
      <button
        class="button button-orange"
        type="button"
      >
        -
      </button>
      <button
        class="button"
        type="button"
      >
        1
      </button>
      <button
        class="button"
        type="button"
      >
        2
      </button>
      <button
        class="button"
        type="button"
      >
        3
      </button>
      <button
        class="button button-orange"
        type="button"
      >
        +
      </button>
      <button
        class="button"
        id="button-0"
        type="button"
      >
        0
      </button>
      <button
        class="button"
        type="button"
      >
        .
      </button>
      <button
        class="button button-orange"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`renders Home component without errors 1`] = `
<div
  class="home-container"
>
  <h1>
    Welcome to Math Magicians App!
  </h1>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec in lacus molestie, dapibus risus in, mollis arcu. Donec eget ultricies lorem. Quisque ut nisi eu massa condimentum pellentesque. Curabitur mauris purus, posuere et ante ac, finibus porta lorem. Duis eget ultricies magna. Aliquam ac orci eu tortor sollicitudin tincidunt. Integer consequat, ante ac pulvinar hendrerit, odio urna pellentesque odio, placerat convallis elit augue nec nulla. Nunc leo purus, auctor at mauris ut, facilisis consequat neque. Aenean ac orci neque. Morbi ac ipsum a erat bibendum rhoncus. Nunc et mattis erat. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Nunc faucibus sollicitudin semper. Donec quis elit sed ipsum dapibus ornare eget sit amet mauris.
  </p>
  <p>
    <br />
    Morbi blandit gravida turpis at congue. Donec vel leo pretium elit gravida eleifend. Duis efficitur turpis lorem, vitae cursus nisl sollicitudin ut. Vestibulum erat tellus, convallis ut est id, volutpat euismod libero. Ut eu tellus a nisi gravida facilisis et et dui. Cras mollis ullamcorper quam sed tempor. Etiam commodo a ante vitae sagittis. In cursus elit nec justo aliquet laoreet. Proin ut mi ipsum. Nullam non leo orci. Etiam pulvinar et nisl non consequat.
  </p>
</div>
`;
